packages/shopping-cart/src/state-based-actions.ts

```diff
@@ -0,0 +1,82 @@
+import debugFactory from 'debug';
+import { playQueuedActions } from './shopping-cart-reducer';
+import type {
+	ShoppingCartState,
+	ShoppingCartReducerDispatch,
+	LastValidResponseCart,
+	ActionPromises,
+	CacheStatus,
+	CartSyncManager,
+	SubscriptionManager,
+} from './types';
+
+const debug = debugFactory( 'shopping-cart:state-based-actions' );
+
+export function prepareFreshCartForInitialFetch(
+	state: ShoppingCartState,
+	dispatch: ShoppingCartReducerDispatch,
+	lastCacheStatus: CacheStatus | ''
+): void {
+	const { cacheStatus } = state;
+	if ( cacheStatus === 'fresh' && cacheStatus !== lastCacheStatus ) {
+		debug( 'triggering fetch of initial cart' );
+		dispatch( { type: 'FETCH_INITIAL_RESPONSE_CART' } );
+	}
+}
+
+function prepareInvalidCartForSync(
+	state: ShoppingCartState,
+	dispatch: ShoppingCartReducerDispatch,
+	lastCacheStatus: CacheStatus | ''
+): void {
+	const { queuedActions, cacheStatus } = state;
+	if (
+		queuedActions.length === 0 &&
+		cacheStatus === 'invalid' &&
+		cacheStatus !== lastCacheStatus
+	) {
+		debug( 'triggering sync of cart to server' );
+		dispatch( { type: 'REQUEST_UPDATED_RESPONSE_CART' } );
+	}
+}
+
+function isStatePendingUpdate( state: ShoppingCartState ) {
+	return state.queuedActions.length > 0 || state.cacheStatus !== 'valid';
+}
+
+export function createTakeActionsBasedOnState(
+	lastValidResponseCart: LastValidResponseCart,
+	actionPromises: ActionPromises,
+	syncManager: CartSyncManager,
+	subscriptionManager: SubscriptionManager
+): ( state: ShoppingCartState, dispatch: ShoppingCartReducerDispatch ) => void {
+	let lastCacheStatus: CacheStatus | '' = '';
+
+	const takeActionsBasedOnState = (
+		state: ShoppingCartState,
+		dispatch: ShoppingCartReducerDispatch
+	) => {
+		const { cacheStatus } = state;
+		debug( 'cache status before state-based-actions is', cacheStatus );
+		prepareFreshCartForInitialFetch( state, dispatch, lastCacheStatus );
+		prepareInvalidCartForSync( state, dispatch, lastCacheStatus );
+
+		syncManager.fetchInitialCartFromServer( state, dispatch );
+		syncManager.syncPendingCartToServer( state, dispatch );
+
+		if ( ! isStatePendingUpdate( state ) ) {
+			debug( 'updating lastValidResponseCart and resolving action promises' );
+			lastValidResponseCart.update( state.responseCart );
+			actionPromises.resolve( state.responseCart );
+		}
+
+		playQueuedActions( state, dispatch );
+
+		lastCacheStatus = cacheStatus;
+
+		subscriptionManager.notifySubscribers();
```

Comment:
> Some actions could easily be synchronous here:

I tried this first, actually, and it didn't work because there's a state where the cart is `valid` and there are no queued actions but we don't actually want to update the response cart or resolve the action promises: just after queued actions are dispatched. I think I resolved the issue now though by making sure the action queue is cleared _after_ the queued actions are dispatched. This was not possible before because each dispatch of the dispatch queue had to be self-contained and so we needed to mark the queue as clear before we dispatched any new actions which otherwise could have triggered duplicate replays; now that dispatch is synchronous, it should be safe.